### PR TITLE
[RFC] matchparen: delayed/lazy display via timers

### PR DESCRIPTION
**Pull request by blueyed**
This improves performance when moving around without really stopping
(i.e. holding h/j/k/l etc).

TODO:
- [x] documentation for the timer variables
- [x] squash?!

### CODE REVIEW

File: runtime/plugin/matchparen.vim [hunk 1]
```diff
@@ -18,14 +18,27 @@ if !exists("g:matchparen_insert_timeout")
   let g:matchparen_insert_timeout = 60
 endif
 
+if !exists('g:matchparen_show_delay')
+  let g:matchparen_show_delay = 100
+endif
+if !exists('g:matchparen_hide_delay')
```
Comment (dpelle) [inline runtime/plugin/matchparen.vim]: We should document these configuration variables.
File runtime/doc/pi_paren.txt looks like the right place.

**Comment by brammool** [comment]
This change should only take effect if the timers feature is supported.

File: runtime/plugin/matchparen.vim [hunk 2]
```diff
@@ -19,18 +20,34 @@ if !exists("g:matchparen_insert_timeout")
 endif
 
 if !exists('g:matchparen_show_delay')
-  let g:matchparen_show_delay = 100
+  let g:matchparen_show_delay = 200
```
Comment (chrisbra) [inline runtime/plugin/matchparen.vim]: why is this increased?

**Comment by chrisbra** [comment]
there is still missing a condition like `if has("timers")` around the timer_start/stop calls

**Comment by kiryph** [comment]
I've downloaded this WIP version of `plugin/matchparen.vim`. However, I've noticed following error message when invoking `:NoMatchParen` and again `:DoMatchParen`:

```
Error detected while processing CursorMoved Auto commands for "*":
E716: Key not present in Dictionary: highlight
```

**Comment by blueyed** [comment]
@kiryph 
Pushed a fixup.

**Comment by kiryph** [comment]
Thanks for this quick fix.

**Comment by kiryph** [comment]
I have one question:

I would like to enable this plugin on demand. So I would like to disable it on startup. I can't use `let g:loaded_matchparen = 1` in my vimrc because the command `:DoMatchParen` will not be defined and I can't activate it on demand.

Is there another solution? Should I add something like `execute("NoMatchParen")` to my vimrc?

**Comment by justinmk** [comment]
@kiryph The purpose of adding an "idle" timer is to eliminate the cost of enabling the plugin. So why make it on-demand?

**Comment by kiryph** [comment]
First of all: right now the timer feature is still WIP. I do not know when this will be merged into upstream. Of course I can maintain my own copy. In general I try to avoid keeping too many things personalized. I prefer sticking to upstream if possible.

Second some people might not like the (delayed) visual feedback and consider it as a distraction, in particular, if there is a time gap. However, sometimes it can be useful and would like to enable it on demand as I was saying. This doesn't seem to be a very complicated feature.

And if one gets really ambitious any loaded plugin contributes to startuptime.

Is this a good enough motivation?

**Comment by blueyed** [comment]
> I do not know when this will be merged into upstream

I guess I just have to address the feedback that has been provided already, but it's not high prio for me: I think it's fine to have this here WIP, so more feedback like yours can be provided.

> some people might not like the (delayed) visual feedback and consider it as a distraction

The timeout would become configurable, and then you set it to a very low number - even when not wanting it to be delayed really, you still not want to have it displayed before e.g. 10ms / all the time really.

Regarding the lazy-loading: most plugin managers allow to trigger lazy-loading in case of commands, so that would work if it would be a separate plugin (and not being loaded anyway by default).
I could imagine configuring the timeout to be -1 would skip enabling it, but still enable/define the commands?!
Or setting up the commands could be done at the top, before `loaded` is checked?

**Comment by kiryph** [comment]
@blueyed Thanks for your feedback. Highly appreciated.

Another thought: does the new version only work if `timers` are available or is there a fallback to the old behaviour?

> I could imagine configuring the timeout to be -1 would skip enabling it, but still enable/define the commands?!
Or setting up the commands could be done at the top, before loaded is checked?

~~I guess I would prefer the first option. I think the variable `g:loaded_matchparen` has a certain semantic meaning. Moving parts of the plugin before would change it. I guess I am a little too petty here. I trust your judgement.~~
I am not sure anymore. I guess moving them beforehand would also make sense since the commands are actually handling enabling/disabling this plugin. Sorry for this clutter.

You should also update the documentation of `runtime/doc/pi_paren.txt`. Afaik this is mandatory before this gets merged.

BTW I really like this change of matchparen.vim. Thanks for this.

**Comment by kiryph** [comment]
A simple solution to activate `pi_paren.vim` on demand could be done as following.
Put this into your vimrc
```viml
autocmd VimEnter * NoMatchParen
" Unimpaired-style mappings
nnoremap [op :DoMatchParen<CR>
nnoremap ]op :NoMatchParen<CR>
nnoremap yop :<C-R>=exists("g:loaded_matchparen") ? 'NoMatchParen' : 'DoMatchParen'<CR><CR>
```
Use `yop` to enable and disable it.

**Comment by blueyed** [comment]
Updated to handle missing timers, and some other improvements.

If you're testing/using this already, please update and report back how it works for you.

**Comment by chrisbra** [comment]
we still need the documentation for the timer variables. Also can you please squash into a single commit? I find it easier to review then.

**Comment by blueyed** [comment]
@chrisbra 
1. Added a TODO list
2. You can view the whole diff at https://patch-diff.githubusercontent.com/raw/vim/vim/pull/1338 and https://patch-diff.githubusercontent.com/raw/vim/vim/pull/1338.diff in raw.
I've thought that it would get squashed / applied manually anyway in the end.

**Comment by blueyed** [comment]
Added docs (a while ago).

**Comment by brammool** [comment]
I tried it, but I don't like the delay.  I also don't like that the highlight disappears after half a second.
I am also not aware of the current implementation causing slowness, thus I don't see a good reason to include this.

**Pull request closed without merging**